packages/multi-test/src/app.rs

```diff
@@ -269,65 +312,107 @@ where
-                let mut events = res.events;
-                // recurse in all messages
-                for resend in res.messages {
-                    let subres = self.execute(resender.clone(), resend)?;
-                    // ignore the data now, just like in wasmd
-                    // append the events
-                    // TODO: is this correct for attributes??? Or do we turn them into sub-events?
-                    attributes.extend_from_slice(&subres.attributes);
-                    events.extend_from_slice(&subres.events);
+
+        // call reply if meaningful
+        if let Ok(r) = res {
+            if matches!(msg.reply_on, ReplyOn::Always | ReplyOn::Success) {
+                let mut orig = r.clone();
+                let reply = Reply {
+                    id: msg.id,
+                    result: ContractResult::Ok(SubMsgExecutionResponse {
+                        events: r.events,
+                        data: r.data,
+                    }),
+                };
+                // do reply and combine it with the original response
+                let res2 = self._reply(contract, reply)?;
+                // override data if set
+                if let Some(data) = res2.data {
+                    orig.data = Some(data);
                 }
-                Ok(AppResponse {
-                    attributes,
-                    events,
-                    data: res.data,
-                })
+                // append the events
+                orig.events.extend_from_slice(&res2.events);
+                Ok(orig)
+            } else {
+                Ok(r)
             }
-            CosmosMsg::Bank(msg) => {
-                self.bank.execute(sender, msg)?;
-                Ok(AppResponse::default())
+        } else if let Err(e) = res {
+            if matches!(msg.reply_on, ReplyOn::Always | ReplyOn::Error) {
+                let reply = Reply {
+                    id: msg.id,
+                    result: ContractResult::Err(e),
+                };
+                self._reply(contract, reply)
+            } else {
+                Err(e)
             }
-            _ => unimplemented!(),
+        } else {
+            res
         }
     }
 
-    fn sudo(&mut self, contract_addr: Addr, msg: Vec<u8>) -> Result<AppResponse, String> {
-        let res = self.wasm.sudo(contract_addr.clone(), self.router, msg)?;
-        let mut attributes = res.attributes;
-        let mut events = res.events;
+    fn _reply(&mut self, contract: Addr, reply: Reply) -> Result<AppResponse, String> {
+        // TODO: process result better, combine events / data from parent
+        let res = self.wasm.reply(contract.clone(), self.querier, reply)?;
+        self.process_response(contract, res)
+    }
+
+    fn process_response(
+        &mut self,
+        contract: Addr,
+        response: Response<C>,
+    ) -> Result<AppResponse, String> {
+        let mut events = response.events;
```

Comment:
For the custom events, wasmd put `wasm-` prefix, so it seems better to put `wasm-` prefix to all custom events types.

https://github.com/CosmWasm/wasmd/blob/39a967ea38d70d84f6da4e7e3a72ba89b562f525/x/wasm/keeper/events.go#L43

Reply:
Very good point here